components/codec/src/columnar/timestamp.rs

```diff
@@ -11,3 +11,301 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
+use bytes_ext::{Buf, BufMut};
+use common_types::time::Timestamp;
+use log::debug;
+use snafu::{ensure, OptionExt};
+
+use crate::{
+    bits::{Bit, BufferedReader, BufferedWriter},
+    columnar::{
+        DecodeContext, InvalidVersion, NotEnoughDatums, ReadEncode, Result, ValuesDecoder,
+        ValuesDecoderImpl, ValuesEncoder, ValuesEncoderImpl,
+    },
+};
+
+/// We use delta-of-delta algorithm to compress timestamp, for details, please refer to this paper http://www.vldb.org/pvldb/vol8/p1816-teller.pdf
+/// The layout for the timestamp:
+/// ```plaintext
+/// +----------------+-------------------------+-------------------+--------------------------+------------------+------------------+-----------------------+----------------------+
+/// | version(8bits) | first_timestamp(64bits) | control_bit(1bit) | delta_of_delta(1~36bits) | delta_of_delta...| end_mask(36bits) | timestamp_len(32bits) | timestamp(64bits)... |
+/// +----------------+-------------------------+-------------------+--------------------------+------------------+------------------+-----------------------+----------------------+
+
+const ENCODE_VERSION: u8 = 1;
+const NUM_BYTES_ENCODE_VERSION_LEN: usize = 1;
+
+const NUM_BYTES_TIMESTAMP_LEN: usize = 4;
+const NUM_BYTES_TIMESTAMP: usize = 8;
+
+/// END_MARKER is a special bit sequence used to indicate the end of the
+/// stream
+const END_MARKER: u64 = 0b1111_0000_0000_0000_0000_0000_0000_0000_0000;
+/// END_MARKER_LEN is the length, in bits, of END_MARKER
+const NUM_BITS_END_MARKER_LEN: usize = 36;
+
+// Encode the timestamp using the delta-of-delta algorithm
+struct TimestampEncoder {}
+
+struct TimestampDecoder {}
+
+enum DeltaOfDeltaEncodeMasks {
```

Comment:
Could you explain these variants?

Reply:
Add some comments to explain the use of these variants.